tokio-current-thread/src/lib.rs

```diff
@@ -480,29 +480,34 @@ impl<'a, P: Park> Entered<'a, P> {
     ///
     /// The caller is responsible for ensuring that other spawned futures
     /// complete execution.
-    pub fn block_on<F>(&mut self, future: F) -> Result<F::Item, BlockError<F::Error>>
+    pub fn block_on<F>(&mut self, mut future: F) -> F::Output
     where
         F: Future,
     {
-        let mut future = executor::spawn(future);
-        let notify = self.executor.scheduler.notify();
+        // Safety: we shadow the original `future`, so it will never move
+        // again.
+        let mut future = unsafe {
+            Pin::new_unchecked(&mut future)
+        };
+        let waker = self.executor.scheduler.waker();
+        let mut cx = Context::from_waker(&waker);
 
         loop {
             let res = self
                 .executor
                 .borrow()
-                .enter(self.enter, || future.poll_future_notify(&notify, 0));
+                .enter(self.enter, || future.as_mut().poll(&mut cx));
 
             match res {
-                Ok(Async::Ready(e)) => return Ok(e),
-                Err(e) => return Err(BlockError { inner: Some(e) }),
-                Ok(Async::NotReady) => {}
+                Poll::Ready(e) => return e,
+                Poll::Pending => {}
             }
 
             self.tick();
 
             if let Err(_) = self.executor.park.park() {
-                return Err(BlockError { inner: None });
+                unimplemented!()
```

Comment:
I'm not certain how best to handle this. Should the park error be returned? That'd frustratingly make `block_on` return a `Result`, when it otherwise seems it wouldn't need to...

Reply:
I think a panic is fine here (documented) and a `try_block_on` can be added later if needed.